README
GNetworkMonitor for Connection Manager
**************************************

Copyright (C) 2012  Intel Corporation. All rights reserved.


This is a GNetworkMonitor plugin that uses network
status information from ConnMan to determine if
the system is in connected state or not.


ChangeLog
ver 0.1:
	Initial release.


test/network-monitor.c
/*
 *
 *  Network Monitor for Connection Manager
 *
 *  Copyright (C) 2012  Intel Corporation. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <stdlib.h>
#include <signal.h>
#include <sys/signalfd.h>

#include <gio/gio.h>

static unsigned int __terminated = 0;
static GMainLoop *loop;

static gboolean check_host(gpointer data)
{
	GNetworkMonitor *monitor;
	GSocketConnectable *addr;
	GCancellable *cancellable;
	GError *error = NULL;
	const char *host = data;
	gboolean reachable;

	monitor =  g_network_monitor_get_default();

	addr = g_network_address_new(host, 80);

	cancellable = g_cancellable_new();

	reachable = g_network_monitor_can_reach(monitor,
						addr,
						cancellable,
						&error);

	g_print("Host %s is %sreachable\n", host, reachable ? "" : "not ");
	if (error != NULL) {
		g_print("Error: %s (%d)\n", error->message, error->code);
		g_error_free(error);
	}

	return FALSE;
}

static void
watch_network_changed(GNetworkMonitor * monitor,
		      gboolean available, gpointer user_data)
{
	g_print("Network is %s\n", available ? "up" : "down");

	g_timeout_add_seconds(0, check_host, user_data);
}

static gboolean signal_handler(GIOChannel *channel, GIOCondition cond,
							gpointer user_data)
{
	struct signalfd_siginfo si;
	ssize_t result;
	int fd;

	if (cond & (G_IO_NVAL | G_IO_ERR | G_IO_HUP))
		return FALSE;

	fd = g_io_channel_unix_get_fd(channel);

	result = read(fd, &si, sizeof(si));
	if (result != sizeof(si))
		return FALSE;

	switch (si.ssi_signo) {
	case SIGINT:
	case SIGTERM:
		if (__terminated == 0) {
			g_print("Terminating");
			g_main_loop_quit(loop);
		}

		__terminated = 1;
		break;
	}

	return TRUE;
}

static guint setup_signalfd(void)
{
	GIOChannel *channel;
	guint source;
	sigset_t mask;
	int fd;

	sigemptyset(&mask);
	sigaddset(&mask, SIGINT);
	sigaddset(&mask, SIGTERM);

	if (sigprocmask(SIG_BLOCK, &mask, NULL) < 0) {
		perror("Failed to set signal mask");
		return 0;
	}

	fd = signalfd(-1, &mask, 0);
	if (fd < 0) {
		perror("Failed to create signal descriptor");
		return 0;
	}

	channel = g_io_channel_unix_new(fd);

	g_io_channel_set_close_on_unref(channel, TRUE);
	g_io_channel_set_encoding(channel, NULL, NULL);
	g_io_channel_set_buffered(channel, FALSE);

	source = g_io_add_watch(channel,
				G_IO_IN | G_IO_HUP | G_IO_ERR | G_IO_NVAL,
				signal_handler, NULL);

	g_io_channel_unref(channel);

	return source;
}

int main(int argc, char **argv)
{
	GNetworkMonitor *monitor;
	gboolean available;
	const char *host = "www.connman.net";
	guint signal;

	g_type_init();

	if (argc > 1 && strcmp(argv[1], "-h") == 0) {
		g_print("Usage: %s [<hostname>]\n", argv[0]);
		exit(-1);
	} else if (argc > 1)
		host = argv[1];

	monitor =  g_network_monitor_get_default();

	g_print("Monitoring via %s\n",
		g_type_name_from_instance((GTypeInstance *) monitor));

	g_signal_connect(monitor, "network-changed",
			G_CALLBACK(watch_network_changed), (gpointer)host);

	available = g_network_monitor_get_network_available(monitor);
	g_print("Initial network availibility is %s\n",
		available ? "yes" : "no");

	signal = setup_signalfd();

	loop = g_main_loop_new(NULL, FALSE);
	g_main_loop_run(loop);

	g_source_remove(signal);

	g_object_unref(monitor);
}


src/connman-api.c
/*
 *
 *  Network Monitor for Connection Manager
 *
 *  Copyright (C) 2012  Intel Corporation. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License version 2.1,
 *  as published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <errno.h>

#include <gio/gio.h>

#include "connman-api.h"

#define CONNMAN_DBUS_NAME "net.connman"
#define CONNMAN_ERROR CONNMAN_DBUS_NAME ".Error"

#define CONNMAN_MANAGER_PATH "/"
#define CONNMAN_MANAGER_INTERFACE CONNMAN_DBUS_NAME ".Manager"

struct connman_manager {
	GDBusConnection *connection;
	guint connman_watch;
	guint property_changed_watch;
	connman_property_changed_cb property_changed_cb;
	void *property_user_data;
	gboolean connman_running;
	GCancellable *pending;
};

static gboolean update_property(struct connman_manager *manager,
				const char *property,
				void *value, void *user_data)
{
	DBG("property %s", property);

	if (manager == NULL || manager->property_changed_cb == NULL)
		return FALSE;

	manager->property_changed_cb(property, value, user_data);

	return TRUE;
}

static void get_properties_callback(GObject *source_object,
				GAsyncResult *res,
				gpointer user_data)
{
	struct connman_manager *manager = user_data;
	GVariant *value, *array, *dictionary;
	GVariantIter iter, tuple;
	GDBusMessage *reply;
	GError *error = NULL;
	const char *key;

	DBG("");

	reply = g_dbus_connection_send_message_with_reply_finish(
					manager->connection, res, &error);
	if (reply == NULL)
		goto error;

	if (error != NULL || g_dbus_message_to_gerror(reply, &error) == TRUE) {
		DBG("%s", error->message);
		g_error_free(error);
		goto error;
	}

	array = g_dbus_message_get_body(reply);
	g_variant_iter_init(&tuple, array);

	dictionary = g_variant_iter_next_value(&tuple);
	g_variant_iter_init (&iter, dictionary);

	while (g_variant_iter_loop(&iter, "{sv}", &key, &value)) {
		if (g_str_equal(key, "State") == TRUE) {
			const char *state;
			gsize len;

			state = g_variant_get_string(value, &len);

			update_property(manager, "State", (void *)state,
					manager->property_user_data);
		}
	}

error:
	if (reply != NULL)
		g_object_unref(reply);

	g_object_unref(manager->pending);
	manager->pending = NULL;
}

static void property_changed_signal_cb(GDBusConnection *connection,
					const gchar *sender_name,
					const gchar *object_path,
					const gchar *interface_name,
					const gchar *signal_name,
					GVariant *parameters,
					gpointer user_data)
{
	struct connman_manager *manager = user_data;
	GVariant *value;
	char *key;

	g_variant_get(parameters, "(sv)", &key, &value);

	if (g_str_equal(key, "State") == TRUE) {
		const char *state;
		gsize len;

		state = g_variant_get_string(value, &len);
		update_property(manager, "State", (void *)state,
				manager->property_user_data);
	}

	g_free(key);
}

static void connman_started(GDBusConnection *conn, const gchar *name,
			const gchar *name_owner, void *user_data)
{
	struct connman_manager *manager = user_data;

	DBG("connection %p manager %p", conn, manager);

	manager->connman_running = TRUE;
}

static void connman_stopped(GDBusConnection *conn, const gchar *name,
			void *user_data)
{
	struct connman_manager *manager = user_data;

	DBG("connection %p manager %p", conn, manager);

	manager->connman_running = FALSE;

	update_property(manager, "State", "idle", manager->property_user_data);
}

struct connman_manager *
connman_manager_init(connman_property_changed_cb property_changed_cb,
			void *user_data)
{
	struct connman_manager *manager;
	GDBusMessage *message;
	int err = 0;

	DBG("");

	manager = g_try_malloc0(sizeof(struct connman_manager));
	if (manager == NULL)
		return NULL;

	manager->connection = g_bus_get_sync(G_BUS_TYPE_SYSTEM, NULL, NULL);
	manager->property_changed_cb = property_changed_cb;
	manager->property_user_data = user_data;

	message = g_dbus_message_new_method_call(CONNMAN_DBUS_NAME,
						CONNMAN_MANAGER_PATH,
						CONNMAN_MANAGER_INTERFACE,
						"GetProperties");
	if (message == NULL) {
		err = -ENOMEM;
		goto error;
	}

	manager->pending = g_cancellable_new();
	g_dbus_connection_send_message_with_reply(manager->connection,
						message,
						G_DBUS_SEND_MESSAGE_FLAGS_NONE,
						-1,
						NULL,
						manager->pending,
						get_properties_callback,
						manager);

	manager->connman_watch = g_bus_watch_name(G_BUS_TYPE_SYSTEM,
						CONNMAN_DBUS_NAME,
						G_BUS_NAME_WATCHER_FLAGS_NONE,
						connman_started,
						connman_stopped,
						manager,
						NULL);
	if (manager->connman_watch == 0) {
		err = -EINVAL;
		goto error;
	}

	manager->property_changed_watch =
		g_dbus_connection_signal_subscribe(manager->connection,
						CONNMAN_DBUS_NAME,
						CONNMAN_MANAGER_INTERFACE,
						"PropertyChanged",
						CONNMAN_MANAGER_PATH,
						NULL,
						G_DBUS_SIGNAL_FLAGS_NONE,
						property_changed_signal_cb,
						manager,
						NULL);

	if (manager->property_changed_watch == 0) {
		err = -EINVAL;
		goto error;
	}

error:
	if (message != NULL)
		g_object_unref(message);

	if (err < 0) {
		DBG("Cannot initialize manager");
		connman_manager_cleanup(manager);
		return NULL;
	}

	return manager;
}

void connman_manager_cleanup(struct connman_manager *manager)
{
	if (manager == NULL)
		return;

	DBG("");

	manager->property_changed_cb = NULL;
	manager->connman_running = FALSE;

	if (manager->pending != NULL)
		g_cancellable_cancel(manager->pending);

	if (manager->property_changed_watch != 0) {
		g_dbus_connection_signal_unsubscribe(manager->connection,
					manager->property_changed_watch);
		manager->property_changed_watch = 0;
	}

	if (manager->connman_watch != 0) {
		g_bus_unwatch_name(manager->connman_watch);
		manager->connman_watch = 0;
	}

	g_object_unref(manager->connection);

	manager->property_user_data = NULL;

	g_free(manager);
	manager = NULL;
}


src/connman-api.h
/*
 *
 *  Network Monitor for Connection Manager
 *
 *  Copyright (C) 2012  Intel Corporation. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License version 2.1,
 *  as published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

typedef void (*connman_property_changed_cb)(const char *property,
					void *value,
					void *user_data);

struct connman_manager;

struct connman_manager *
connman_manager_init(connman_property_changed_cb property_changed_cb,
			void *user_data);

void connman_manager_cleanup(struct connman_manager *manager);

#define DBG(fmt, arg...) do {					       \
	g_debug("%s:%s() " fmt "\n", __FILE__, __FUNCTION__ , ## arg); \
} while (0)


src/connman-network-monitor.c
/*
 *
 *  Network Monitor for Connection Manager
 *
 *  Copyright (C) 2012  Intel Corporation. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License version 2.1,
 *  as published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <glib.h>
#include <gio/gio.h>

#include "connman-api.h"

static int priority = 90;
static guint network_changed_signal = 0;

enum {
	PROP_0,
	PROP_NETWORK_AVAILABLE,
};

enum connman_state {
	STATE_UNKNOWN = 0,
	STATE_IDLE,
	STATE_READY,
	STATE_ONLINE,
	STATE_FAILURE,
};

typedef struct _GNetworkMonitorConnmanPrivate GNetworkMonitorConnmanPrivate;
struct _GNetworkMonitorConnmanPrivate
{
	enum connman_state state;
	struct connman_manager *manager;
};

typedef struct _GNetworkMonitorConnman GNetworkMonitorConnman;
struct _GNetworkMonitorConnman {
	GObject parent;
	GNetworkMonitorConnmanPrivate *priv;
};

typedef struct _GNetworkMonitorConnmanClass GNetworkMonitorConnmanClass;
struct _GNetworkMonitorConnmanClass {
	GObjectClass parent_class;
};

#define CONNMAN_TYPE_NETWORK_MONITOR (g_network_monitor_connman_get_type())
#define CONNMAN_NETWORK_MONITOR(obj) (G_TYPE_CHECK_INSTANCE_CAST((obj),	\
					CONNMAN_TYPE_NETWORK_MONITOR,   \
					GNetworkMonitorConnman))
#define CONNMAN_NETWORK_MONITOR_CLASS(klass) \
					(G_TYPE_CHECK_CLASS_CAST((klass), \
						CONNMAN_TYPE_NETWORK_MONITOR, \
						GNetworkMonitorConnmanClass))
#define CONNMAN_IS_NETWORK_MONITOR(obj) (G_TYPE_CHECK_INSTANCE_TYPE((obj), \
						CONNMAN_TYPE_NETWORK_MONITOR))
#define CONNMAN_IS_NETWORK_MONITOR_CLASS(klass) \
					(G_TYPE_CHECK_CLASS_TYPE((klass), \
						CONNMAN_TYPE_NETWORK_MONITOR))

static void network_monitor_iface_init(GNetworkMonitorInterface *iface);
static void network_monitor_initable_iface_init(GInitableIface *iface);
static void g_network_monitor_connman_init(GNetworkMonitorConnman *self);
static void g_network_monitor_connman_class_init(GNetworkMonitorConnmanClass *klass);
GType g_network_monitor_connman_get_type(void);

G_DEFINE_TYPE_WITH_CODE(GNetworkMonitorConnman,
			g_network_monitor_connman,
			G_TYPE_OBJECT,
			G_IMPLEMENT_INTERFACE(G_TYPE_INITABLE,
				network_monitor_initable_iface_init)
			G_IMPLEMENT_INTERFACE(G_TYPE_NETWORK_MONITOR,
				network_monitor_iface_init)
			g_io_extension_point_implement(
				G_NETWORK_MONITOR_EXTENSION_POINT_NAME,
				g_define_type_id,
				"connman",
				priority))

static GObject *network_monitor_constructor(GType type, guint n_props,
						GObjectConstructParam *props)
{
	GObject *obj;

	obj = G_OBJECT_CLASS(g_network_monitor_connman_parent_class)->
					constructor(type, n_props, props);
	return obj;
}

static void network_monitor_finalize(GObject *object)
{
	GNetworkMonitorConnman *monitor;

	g_return_if_fail(object != NULL);
	g_return_if_fail(CONNMAN_IS_NETWORK_MONITOR(object));

	monitor = CONNMAN_NETWORK_MONITOR(object);

	connman_manager_cleanup(monitor->priv->manager);
	monitor->priv->manager = NULL;
	monitor->priv->state = STATE_UNKNOWN;

	G_OBJECT_CLASS(g_network_monitor_connman_parent_class)->
					finalize(object);
}

static gboolean is_available(enum connman_state state)
{
	switch (state) {
	case STATE_UNKNOWN:
	case STATE_IDLE:
	case STATE_FAILURE:
		return FALSE;

	case STATE_READY:
	case STATE_ONLINE:
		return TRUE;
	}

	return FALSE;
}

static gboolean get_state(GNetworkMonitorConnman *monitor)
{
	return is_available(monitor->priv->state);
}

static void get_property(GObject *object, guint prop_id,
			GValue *value, GParamSpec *pspec)
{
	GNetworkMonitorConnman *monitor = CONNMAN_NETWORK_MONITOR(object);

	switch (prop_id) {
	case PROP_NETWORK_AVAILABLE:
		g_value_set_boolean(value, get_state(monitor));
		break;

	default:
		G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
		break;
	}
}

static void
g_network_monitor_connman_class_init(GNetworkMonitorConnmanClass *klass)
{
	GObjectClass *gobject_class;

	gobject_class = G_OBJECT_CLASS(klass);
	gobject_class->constructor = network_monitor_constructor;
	gobject_class->finalize = network_monitor_finalize;
	gobject_class->get_property = get_property;

	g_object_class_override_property(gobject_class,
					PROP_NETWORK_AVAILABLE,
					"network-available");

	g_type_class_add_private(gobject_class,
				sizeof(GNetworkMonitorConnmanPrivate));
}

static enum connman_state string2state(const char *state)
{
	if (g_strcmp0(state, "idle") == 0)
		return STATE_IDLE;
	else if (g_strcmp0(state, "ready") == 0)
		return STATE_READY;
	else if (g_strcmp0(state, "online") == 0)
		return STATE_ONLINE;
	else if (g_strcmp0(state, "failure") == 0)
		return STATE_FAILURE;

	return STATE_UNKNOWN;
}

static void property_changed(const char *property, void *value,
							void *user_data)
{
	GNetworkMonitorConnman *monitor = user_data;
	enum connman_state old_state, new_state;

	if (monitor == NULL) {
		DBG("monitor missing");
		return;
	}

	old_state = new_state = monitor->priv->state;

	if (g_strcmp0(property, "State") == 0) {
		new_state = string2state(value);
		DBG("property %s value \"%s\"", property, (char *)value);
	}

	if (is_available(new_state) != is_available(old_state)) {
		monitor->priv->state = new_state;
		g_signal_emit(monitor, network_changed_signal, 0,
							get_state(monitor));
	}
}

static void g_network_monitor_connman_init(GNetworkMonitorConnman *self)
{
	self->priv = G_TYPE_INSTANCE_GET_PRIVATE(self,
						CONNMAN_TYPE_NETWORK_MONITOR,
						GNetworkMonitorConnmanPrivate);
}

static gboolean network_monitor_initable_init(GInitable *initable,
					GCancellable *cancellable,
					GError **error)
{
	GNetworkMonitorConnman *cm = CONNMAN_NETWORK_MONITOR(initable);

	cm->priv->manager = connman_manager_init(property_changed, cm);

	DBG("cm %p manager %p", cm, cm->priv->manager);

	return TRUE;
}

static void network_monitor_initable_iface_init(GInitableIface *iface)
{
	iface->init = network_monitor_initable_init;
}

static gboolean is_local(GSocketConnectable *connectable,
			GCancellable *cancellable)
{
	GSocketAddressEnumerator *enumerator;
	GSocketAddress *addr;
	GError *error = NULL;

	enumerator = g_socket_connectable_proxy_enumerate(connectable);
	addr = g_socket_address_enumerator_next(enumerator, cancellable,
						&error);
	if (addr == NULL) {
		g_object_unref(enumerator);
		return FALSE;
	}

	while (addr != NULL) {
		if (G_IS_INET_SOCKET_ADDRESS(addr)) {
			GInetAddress *iaddr;

			iaddr = g_inet_socket_address_get_address(
						G_INET_SOCKET_ADDRESS(addr));
			if (g_inet_address_get_is_loopback(iaddr) == TRUE) {
				g_object_unref(addr);
				g_object_unref(enumerator);
				return TRUE;
			}
		}

		g_object_unref(addr);
		addr = g_socket_address_enumerator_next(enumerator,
							cancellable,
							&error);
	}

	g_object_unref(enumerator);
	return FALSE;
}

static gboolean can_reach(GNetworkMonitor *monitor,
				GSocketConnectable *connectable,
				GCancellable *cancellable,
				GError **error)
{
	GNetworkMonitorConnman *cm = CONNMAN_NETWORK_MONITOR(monitor);

	DBG("");

	if (get_state(cm) == FALSE) {
		if (is_local(connectable, cancellable) == TRUE)
			return TRUE;

		if (error && *error == NULL)
			g_set_error_literal(error, G_IO_ERROR,
					G_IO_ERROR_NETWORK_UNREACHABLE,
					"No network connections available");
		return FALSE;
	}

	return TRUE;
}

static void network_monitor_iface_init(GNetworkMonitorInterface *iface)
{
	network_changed_signal = g_signal_lookup("network-changed",
						G_TYPE_NETWORK_MONITOR);

	iface->can_reach = can_reach;
}

void g_io_module_load(GIOModule *module)
{
	GIOExtensionPoint *ep;

	ep = g_io_extension_point_register(G_NETWORK_MONITOR_EXTENSION_POINT_NAME);
	if (ep == NULL) {
		g_warning("Extension point %s not found",
			G_NETWORK_MONITOR_EXTENSION_POINT_NAME);
		return;
	}

	g_io_extension_point_set_required_type(ep,
					CONNMAN_TYPE_NETWORK_MONITOR);

	g_type_module_use(G_TYPE_MODULE(module));
}

void g_io_module_unload(GIOModule *module)
{
}

char** g_io_module_query(void)
{
	char **eps = g_try_new(char *, 2);
	eps[0] = g_strdup(G_NETWORK_MONITOR_EXTENSION_POINT_NAME);
	eps[1] = NULL;
	return eps;
}
